### PR TITLE
Caplin: Introduced cached merkle tree

### PR DESCRIPTION
**Pull request by Giulio2002**
# Merkle Tree cache for Intermediate hashes

this PR implements a cached Merkle tree for ARM-based computers which do not have significant gains from `gohashtree`.

### Merkle Tree Structure

The `MerkleTree` struct contains:
- `computeLeaf`: A function to compute the hash of a leaf node.
- `layers`: A slice of byte slices representing the intermediate layers of the tree.
- `leavesCount`: The number of leaves in the tree.
- `hashBuf`: A buffer to store input for hashing.
- `limit`: An optional limit for the number of leaves.

### Key Constants

- `OptimalMaxTreeCacheDepth`: A constant defining the maximum depth of the tree cache.

### Initialization

The `Initialize` method sets up the Merkle tree with a specified number of leaves and a maximum cache depth. It also initializes the layers and sets an optional limit for the number of leaves.

### Marking Leaves as Dirty

The `MarkLeafAsDirty` method marks a leaf node as "dirty", indicating it needs to be recomputed during the next root calculation.

### Adding Leaves

The `AppendLeaf` method allows adding a new leaf to the tree. It marks the new leaf as dirty and extends the layers as needed.

### Extending Layers

The `extendLayer` method extends a given layer by 1.5 times its current size, ensuring the new leaf is marked as dirty.

### Computing the Root

The `ComputeRoot` method calculates the root hash of the Merkle tree. It handles various cases:
- No layers or leaves: Returns a predefined zero hash.
- Few leaves: Computes the root directly from the leaves.
- Larger trees: Computes intermediate layers to derive the root hash.

### Copying the Tree

The `CopyInto` method copies the current Merkle tree into another instance.

### Internal Helper Methods

- `finishHashing`: Completes the hashing process for the last layer.
- `computeLayer`: Computes the hash for a specific layer.

## High-Level Workflow

1. **Initialization**: Set up the tree with the desired number of leaves and depth.
2. **Mark Leaves**: Mark leaves as dirty when they need to be recomputed.
3. **Append Leaves**: Add new leaves to the tree, extending layers as necessary.
4. **Compute Root**: Calculate the root hash of the tree, which can be used to verify the integrity of all leaves.
5. **Copy Tree**: Create a copy of the tree if needed.

## Intermediate Hash Storage

The implementation stores only intermediate hashes, not the leaves themselves. This is achieved through the following:

- **Intermediate Layers**: The `layers` field holds intermediate hashes between the leaves and the root. Each entry in `layers` represents a level in the Merkle tree, starting from the first intermediate layer up to the layer just before the root.
- **Leaf Calculation**: The `computeLeaf` function is used to compute the hash of a leaf node when needed, but these leaf hashes are not stored directly in the `layers`.
- **Root Calculation**: When computing the root, the tree uses the intermediate hashes stored in `layers` to derive the final root hash. This ensures that only the necessary intermediate hashes are kept in memory, optimizing storage efficiency.

## Rationale for Intermediate Hash Storage

The primary rationale for storing only intermediate hashes and not recomputing leaves is efficiency. By retaining intermediate hashes:
- **Reduced Computation**: The tree avoids recomputing leaf hashes unless absolutely necessary, saving computational resources.
- **Optimized Storage**: Only the hashes essential for reconstructing the root are stored, minimizing memory usage.
- **Performance Gains**: This approach leads to faster root computation, especially in trees with a large number of leaves.

### CODE REVIEW

File: cl/cltypes/solid/validator_set.go [hunk 1]
```diff
@@ -144,10 +138,16 @@ func (v *ValidatorSet) CopyTo(t *ValidatorSet) {
...
+		if t.MerkleTree == nil {
+			t.MerkleTree = &merkle_tree.MerkleTree{}
+		}
+		v.MerkleTree.CopyInto(t.MerkleTree)
+	}
 	// skip copying (unsupported for phase0)
 	t.phase0Data = make([]Phase0Data, t.l)
 	copy(t.buffer, v.buffer)
-	copy(t.treeCacheBuffer, v.treeCacheBuffer)
+	// copy(t.treeCacheBuffer, v.treeCacheBuffer) TODO: Implement CopyTo in MerkleTree
```
Comment (domiwei) [inline cl/cltypes/solid/validator_set.go:150]:
```suggestion
```

File: cl/merkle_tree/merkle_tree.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+		}
+	}
+}
+
+func (m *MerkleTree) AppendLeaf() {
+	/*
+		Step 1: Append a new dirty leaf
+		Step 2: Extend each layer with the new leaf when needed (1.5x extension)
+	*/
+	for i := 0; i < len(m.layers); i++ {
```
Comment (domiwei) [inline cl/merkle_tree/merkle_tree.go:68]: I am not sure what will happen if:
1. initialize with leaf size 2
2. AppendLeaf 3 times with non-empty leaf (i know which is fetched in m.computeLeaf(...))
3. then compute root

so in this case, the root layer turns from Layer0 to Layer2,  but this loop seems not extend new layers?
or, extend new layers while computing root in Line154? https://github.com/erigontech/erigon/pull/11096/files?diff=unified&w=0#diff-5c053bedbcbe2473e56d2a0b317ba9e2f4d6f169faaf83e6b5bdd463cc2b7bceR154 
ignore me if this case can be properly handled

**Comment by Giulio2002** [comment]
I will address comments and all after the Erigon 3.0 release

File: cl/merkle_tree/merkle_tree.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+				m.computeLeaf(leafIndexBegin+1, m.hashBuf[length.Hash:])
+			}
+			if err := HashByteSlice(m.layers[layerIdx][fromOffset:toOffset], m.hashBuf[:]); err != nil {
+				panic(err)
+			}
+			continue
+		}
+		childFromOffset := (i * 2) * length.Hash
+		childToOffset := (i*2 + 2) * length.Hash
+		if childToOffset > len(m.layers[layerIdx-1]) {
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply cl/merkle_tree/merkle_tree.go:239]: impossible case

File: cl/merkle_tree/merkle_tree.go (hunk 2 above)
Comment (Giulio2002) [reply cl/merkle_tree/merkle_tree.go:68]: It is handled:


## Initial State

```
        root
leaf_1  leaf_2
```

## Adding Two Leafs

When adding a new leaf node, the following thing occur:

```
                               root (extended)
                              /                 \
            intermediate_node_1       intermediate_node_2 (newly added)
           /        \                   /           \
      leaf_1     leaf_2            leaf_3     empty_leaf (newly added)
```

## Steps Explained

- **Initial State**: 
    - The tree starts with a `root` and two leaves: `leaf_1` and `leaf_2`.

- **First Extension**:
    - When extending the tree, an intermediate layer is appended to the root.
    - `intermediate_node_1` and `intermediate_node_2` are introduced.
    - `leaf_3` and an `empty_leaf` are added as children of `intermediate_node_2`.

The structure now has an extended root with intermediate nodes and new leaf nodes.

File: cl/merkle_tree/merkle_tree.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+		}
+		newLayerSize := newExpectendLayerNodeCount * length.Hash
+		capacity := (newLayerSize / 2) * 3
+		m.layers[layerIdx] = make([]byte, newLayerSize, capacity)
+	}
+	if len(m.layers[layerIdx]) == 0 {
+		return
+	}
+	iterations := m.leavesCount / currentDivisor
+	if m.leavesCount%currentDivisor != 0 {
```
[the remark this answers is not in the document]
Comment (Giulio2002) [reply cl/merkle_tree/merkle_tree.go:213]: good idea, however, in reality most of those ceil operations do include the operations but needs the check that the tree does not end there. but good for this case

**Comment by Giulio2002** [comment]
@domiwei all done, review again.

File: cl/merkle_tree/merkle_tree.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (domiwei) [reply cl/merkle_tree/merkle_tree.go:241]: sorry for this stupid question...

**Review by domiwei — APPROVED** [review]

**Comment by domiwei** [comment]
> @domiwei all done, review again.

lgtm, but i noticed that seems this cached merkle tree struct is not thread-safe. Maybe you need mutex to protect it?

**Comment by Giulio2002** [comment]
> > @domiwei all done, review again.
> 
> lgtm, but i noticed that seems this cached merkle tree struct is not thread-safe. Maybe you need mutex to protect it?

No issue, not supposed to

**Pull request merged**